sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UseUriInsteadOfString.cs

```diff
@@ -51,75 +53,143 @@ public sealed class UseUriInsteadOfString : SonarDiagnosticAnalyzer
+        private static readonly DiagnosticDescriptor rule_S4005 =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId_RuleS4005, MessageFormat_RuleS4005, RspecStrings.ResourceManager);
         #endregion
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
-            => ImmutableArray.Create(rule_S3994, rule_S3995, rule_S3996);
+            => ImmutableArray.Create(rule_S3994, rule_S3995, rule_S3996, rule_S3997, rule_S4005);
 
         private static readonly HashSet<string> UrlNameVariants = new HashSet<string> { "uri", "url", "urn" };
 
         protected sealed override void Initialize(SonarAnalysisContext context)
         {
+            context.RegisterSyntaxNodeActionInNonGenerated(VerifyDeclaration,
+                SyntaxKind.MethodDeclaration, SyntaxKind.ConstructorDeclaration);
+
             context.RegisterSyntaxNodeActionInNonGenerated(
                 c =>
                 {
-                    var methodDeclaration = (MethodDeclarationSyntax)c.Node;
-                    var methodSymbol = c.SemanticModel.GetDeclaredSymbol(methodDeclaration);
-
-                    if (methodSymbol.IsOverride)
-                    {
-                        return;
-                    }
+                    var propertyDeclaration = (PropertyDeclarationSyntax)c.Node;
+                    var propertySymbol = c.SemanticModel.GetDeclaredSymbol(propertyDeclaration);
 
-                    if (methodSymbol.ReturnType.Is(KnownType.System_String) &&
-                        NameContainsUri(methodSymbol.Name))
+                    if (propertySymbol.Type.Is(KnownType.System_String) &&
+                        !propertySymbol.IsOverride &&
+                        NameContainsUri(propertySymbol.Name))
                     {
-                        c.ReportDiagnostic(Diagnostic.Create(rule_S3995, methodDeclaration.ReturnType.GetLocation()));
+                        c.ReportDiagnostic(Diagnostic.Create(rule_S3996, propertyDeclaration.Type.GetLocation()));
                     }
+                },
+                SyntaxKind.PropertyDeclaration);
 
-                    var stringUrlParams = GetStringUrlParamIndexes(methodSymbol);
-                    if (!stringUrlParams.Any() ||
-                        HasOverloadThatUsesUriTypeInPlaceOfString(methodSymbol, stringUrlParams))
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var invokedMethodSymbol = c.SemanticModel.GetSymbolInfo(c.Node).Symbol as IMethodSymbol;
+                    if (invokedMethodSymbol == null || invokedMethodSymbol.IsInType(KnownType.System_Uri))
                     {
                         return;
                     }
 
-                    foreach (int paramIdx in stringUrlParams)
+                    var stringUrlParams = GetStringUrlParamIndexes(invokedMethodSymbol);
```

Comment:
For some reason I think we should be able to catch this (I am not sure if we do already)

    public void Foo(string url)
    {
        Foo(pez, new Uri(url), bar, baz);
    }